File: frontend/src/types/notes.ts
export interface NoteFolder {
  id: number;
  name: string;
  parent_id: number | null;
  note_count: number;
  children: NoteFolder[];
  created_at: string;
  updated_at: string;
}

export interface NoteFolderCreate {
  name: string;
  parent_id?: number | null;
}

export interface NoteFolderUpdate {
  name?: string;
  parent_id?: number | null;
}

export interface NoteSummary {
  id: number;
  title: string;
  folder_id: number | null;
  created_at: string;
  updated_at: string;
}

export interface Note {
  id: number;
  title: string;
  content: string;
  folder_id: number | null;
  created_at: string;
  updated_at: string;
}

export interface NoteCreate {
  title?: string;
  content?: string;
  folder_id?: number | null;
}

export interface NoteUpdate {
  title?: string;
  content?: string;
  folder_id?: number | null;
}

// Virtual folder sentinel for notes with no folder
export const UNFILED_FOLDER_ID = "unfiled" as const;
export type FolderSelection = number | typeof UNFILED_FOLDER_ID | null; // null = all notes

File: backend/app/models/notes.py
from datetime import datetime

from sqlalchemy import DateTime, ForeignKey, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class NoteFolder(Base):
    __tablename__ = "note_folders"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    parent_id: Mapped[int | None] = mapped_column(
        Integer, ForeignKey("note_folders.id", ondelete="SET NULL"), nullable=True, index=True
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False
    )

    parent: Mapped["NoteFolder | None"] = relationship(
        "NoteFolder", remote_side="NoteFolder.id", back_populates="children"
    )
    # No cascade — we re-parent children manually on delete
    children: Mapped[list["NoteFolder"]] = relationship(
        "NoteFolder", back_populates="parent"
    )
    notes: Mapped[list["Note"]] = relationship(
        "Note", back_populates="folder"
    )


class Note(Base):
    __tablename__ = "notes"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    title: Mapped[str] = mapped_column(String(500), nullable=False, default="Untitled")
    content: Mapped[str] = mapped_column(Text, nullable=False, default="")
    folder_id: Mapped[int | None] = mapped_column(
        Integer, ForeignKey("note_folders.id", ondelete="SET NULL"), nullable=True, index=True
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False
    )

    folder: Mapped["NoteFolder | None"] = relationship("NoteFolder", back_populates="notes")
